Hazel/src/Hazel/Renderer/Shader.cpp

```diff
@@ -30,41 +30,41 @@ namespace Hazel {
 		return nullptr;
 	}
 
-	void ShaderLibrary::Add(const std::string& name, const Ref<Shader>& shader)
+	void ShaderLibrary::Add(std::string_view name, const Ref<Shader>& shader)
 	{
 		HZ_CORE_ASSERT(!Exists(name), "Shader already exists!");
-		m_Shaders[name] = shader;
+		m_Shaders[name.data()] = shader;
 	}
 
 	void ShaderLibrary::Add(const Ref<Shader>& shader)
 	{
-		auto& name = shader->GetName();
+		auto name = shader->GetName();
 		Add(name, shader);
 	}
 
-	Ref<Shader> ShaderLibrary::Load(const std::string& filepath)
+	Ref<Shader> ShaderLibrary::Load(std::string_view filepath)
 	{
 		auto shader = Shader::Create(filepath);
 		Add(shader);
 		return shader;
 	}
 
-	Ref<Shader> ShaderLibrary::Load(const std::string& name, const std::string& filepath)
+	Ref<Shader> ShaderLibrary::Load(std::string_view name, std::string_view filepath)
 	{
 		auto shader = Shader::Create(filepath);
 		Add(name, shader);
 		return shader;
 	}
 
-	Ref<Shader> ShaderLibrary::Get(const std::string& name)
+	Ref<Shader> ShaderLibrary::Get(std::string_view name)
 	{
 		HZ_CORE_ASSERT(Exists(name), "Shader not found!");
-		return m_Shaders[name];
+		return m_Shaders[name.data()];
```

Comment:
You shouldn't be using `std::string_view::data` as an accessor to a null-terminated byte string, because it isn't.

This would now be broken with this change
```
std::string_view v = "hello world";
shaderLib.Get(v.substr(5));
```

Looking at this review, much of the `string_view` changes also have broken this, not to mention in some cases it is actually adding unnecessary conversions between `std::string` and `std::string_view` which will just create additional allocations and overhead such as here because `unordered_map::operator[]` take the template `Key` type not some other template type (`find` did get this in C++20 but not `operator[]`).

I feel it is reckless to make these changes in bulk without considering all of the use cases, moving straight to `string_view` does not mean you instantly get performance benefits, if performance is a concern with a `string` the solution of going straight to `string_view` is often not the best solution, strings are slow, error prone, etc.

All that moving it to a `string_view` has done is remove the edge case of a string literal constructing a string each time, while making it harder to work many API that expect null-terminated strings (e.g. OpenGL), even a `const char *` is better in many cases then a `string_view`

Reply:
Using string_view in class member initialization and a couple of other places is not recommended in C++17 book by Nicolai M. Josuttis "C++17 - The Complete Guide" ; he gives summary of safe use of string_views ;

" Summary of Safe Use of String Views
To summarize, use std::string_view with care, meaning that you should adjust your programming style as follows:
• Do not use string views in APIs that pass the argument to a string.
– Do not initialize string members from string_view parameters.
– No string at the end of a string_ view chain.
• Do not return a string view."

class Person {
  std::string name;
public:
Person (std::string_view n)
: name{n} { }
// don’t do this

Author recommended using string as a copy and then std::move() is a better to initialize the
class Person {
     std::string name;
    public:
     Person (std::string n)
      : name{std::move(n)} {
     }
...
};